sklearn/mixture/_gaussian_mixture_ic.py

```diff
@@ -0,0 +1,813 @@
+
+    def _fit_cluster(
+        self, X, X_subset, y, ag_params, gm_params, agg_clustering, seed
+    ):
+        label_init = self.label_init
+        if label_init is not None:
+            gm_params = self._init_gm_params(X, label_init, gm_params)
+        elif ag_params["affinity"] != "none":
+            gm_params = self._init_gm_params(
+                X_subset, agg_clustering, gm_params
+            )
+        else:
+            gm_params["init_params"] = "kmeans"
+        gm_params["reg_covar"] = 0
+        gm_params["max_iter"] = self.max_iter
+        gm_params["random_state"] = seed
+
+        # if none of the iterations converge, bic/aic is set to inf
+        criter = np.inf
+        # below is the regularization scheme
+        while gm_params["reg_covar"] <= 1 and criter == np.inf:
+            model = GaussianMixture(**gm_params)
+            try:
+                # ignoring warning here because if convergence is not reached,
+                # the regularization is automatically increased
+                with warnings.catch_warnings():
+                    warnings.simplefilter("ignore", ConvergenceWarning)
+                    model.fit(X)
+                predictions = model.predict(X)
+                counts = [
+                    sum(predictions == i)
+                    for i in range(gm_params["n_components"])
+                ]
+                # singleton clusters not allowed
+                assert not any([count <= 1 for count in counts])
+
+            except ValueError:
+                gm_params["reg_covar"] = _increase_reg(gm_params["reg_covar"])
+                continue
+            except AssertionError:
+                gm_params["reg_covar"] = _increase_reg(gm_params["reg_covar"])
+                continue
+            # if the code gets here, then the model has been fit with
+            # no errors or singleton clusters
+            if self.criterion == "bic":
+                criter = model.bic(X)
+            else:
+                criter = model.aic(X)
+            break
+
+        results = _CollectResults(model, criter, gm_params, ag_params)
+        return results
+
+    def fit(self, X, y=None):
+        """
+        Fit several Gaussian mixture models to the data.
+        Initialize with agglomerative clustering then
+        estimate model parameters with EM algorithm.
+        Select the best model according to the chosen
+        information criterion.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            List of n_features-dimensional data points. Each row
+            corresponds to a single data point.
+
+        Returns
+        -------
+        self : object
+            Returns an instance of self.
+        """
+
+        self._check_parameters()
+        X = check_array(
+            X, dtype=[np.float64, np.float32], ensure_min_samples=1
+        )
+
+        random_state = check_random_state(self.random_state)
```

Comment:
im not sure, but does this need to be a `self.random_state = check_random_state(self.random_state)`?